instrumentation/apache-httpclient-4.0/src/main/java/io/opentelemetry/instrumentation/hypertrace/apachehttpclient/v4_0/ApacheHttpClientUtils.java

```diff
@@ -0,0 +1,109 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.hypertrace.apachehttpclient.v4_0;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.trace.Span;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.Charset;
+import java.util.function.Function;
+import org.apache.http.Header;
+import org.apache.http.HeaderIterator;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpEntityEnclosingRequest;
+import org.apache.http.HttpMessage;
+import org.hypertrace.agent.config.Config.AgentConfig;
+import org.hypertrace.agent.core.ContentEncodingUtils;
+import org.hypertrace.agent.core.HypertraceConfig;
+import org.hypertrace.agent.core.HypertraceSemanticAttributes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ApacheHttpClientUtils {
+  private ApacheHttpClientUtils() {}
+
+  private static final Logger log = LoggerFactory.getLogger(ApacheHttpClientUtils.class);
+
+  public static void addResponseHeaders(Span span, HeaderIterator headerIterator) {
+    addHeaders(span, headerIterator, HypertraceSemanticAttributes::httpResponseHeader);
+  }
+
+  public static void addRequestHeaders(Span span, HeaderIterator headerIterator) {
+    addHeaders(span, headerIterator, HypertraceSemanticAttributes::httpRequestHeader);
+  }
+
+  private static void addHeaders(
+      Span span,
+      HeaderIterator headerIterator,
+      Function<String, AttributeKey<String>> attributeKeySupplier) {
+    while (headerIterator.hasNext()) {
+      Header header = headerIterator.nextHeader();
+      span.setAttribute(attributeKeySupplier.apply(header.getName()), header.getValue());
+    }
+  }
+
+  public static void traceRequest(HttpMessage request) {
+    Span currentSpan = Span.current();
+    AgentConfig agentConfig = HypertraceConfig.get();
+    if (agentConfig.getDataCapture().getHttpHeaders().getRequest().getValue()) {
+      ApacheHttpClientUtils.addRequestHeaders(currentSpan, request.headerIterator());
+    }
+
+    if (agentConfig.getDataCapture().getHttpBody().getRequest().getValue()
+        && request instanceof HttpEntityEnclosingRequest) {
+      HttpEntityEnclosingRequest entityRequest = (HttpEntityEnclosingRequest) request;
+      HttpEntity entity = entityRequest.getEntity();
+      ApacheHttpClientUtils.traceEntity(
+          currentSpan, HypertraceSemanticAttributes.HTTP_REQUEST_BODY.getKey(), entity);
+    }
+  }
+
+  public static void traceEntity(Span span, String bodyAttributeKey, HttpEntity entity) {
+    if (entity == null) {
+      return;
+    }
+
+    if (entity.isRepeatable()) {
```

Comment:
we won't capture if it isn't repeatable ?

Reply:
see line 107

Reply:
It captures bodies for repeatable but also non-repeatable streams. There is also a unit test to verify it.